src/sentry/search/events/builder/metrics.py

```diff
@@ -86,6 +88,53 @@ def __init__(
             raise InvalidSearchQuery("Organization id required to create a metrics query")
         self.organization_id: int = org_id
 
+        self._on_demand_spec = self._resolve_on_demand_spec(
+            kwargs.get("selected_columns", []), kwargs.get("query", "")
+        )
+
+    def _resolve_on_demand_spec(
+        self, selected_cols: List[Optional[str]], query: str
+    ) -> Optional[OndemandMetricSpec]:
+        field = selected_cols[0] if selected_cols and len(selected_cols) else None
+        if not self.is_performance or not self.is_alerts_query or not field:
+            return None
+
+        try:
+            if OndemandMetricSpec.check(field, query):
+                return OndemandMetricSpec.parse(field, query)
+
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
```

Comment:
With the proposed class impl this would become something like:
```suggestion
        try:
            generator = OnDemandMetricGenerator(field, query)
            return generator.parse() # No need to do the check since it's already built-in.
        except Exception as e:
            sentry_sdk.capture_exception(e)
```

A better implementation would be to throw an exception and catch is here and then return `None` but it's mostly a matter of preference.